### DISCUSSION TITLE
Backpropagating losses for two different outputs

### DISCUSSION THREAD

**Discussion by enter802** [pyg-team/pytorch_geometric#7618] [Q&A] [answered]
I have the network below with two different outputs. To train the network, I tried adding the two losses and backpropagate the total, however, doing that, the performance of `y1` only improves - `y2` remains pretty much stale, in fact, worsening in some cases (different initialization, different hyperparameters). I was wondering if there is any other way(s) the network could be trained. For example, I was wondering if it is possible to train the weights of `MLP1` and `MLP2` with `L1` and `L2` respectively, while training the weights of the `GNN` with `L1+L2`. How do you code in pytorch to do that exactly?

Thanks for your inputs in advance and appreciate any guidance.

<img width="511" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/9440833/9a16e45a-bd8a-4608-82a0-8021bcfc1920">

**Comment by rusty1s** [comment]
How does your loss function looks like? It may well be that you need to tune this quite a bit, e.g., by increasing the L2` loss:
```python
loss = loss_1 + factor * loss_2
loss.backward()

**Reply by enter802** [reply]
My loss function looks the same as you have suggested except that I used `factor=1`. I will try increasing the factor. However, I was hoping if there is a way to implement what I asked earlier:

is it possible to train the weights of `MLP1` and `MLP2` with `L1` and `L2` respectively, while training the weights of the `GNN` with `L1+L2`?

your response is much appreciated. thanks.

**Reply by rusty1s** [reply, marked as answer]
That is what will automatically happen since any gradients from L1 will only go back through MLP1, not MLP2.

**Reply by enter802** [reply]
oh. that makes sense. Thanks a lot.
